### PR TITLE
Add broadcast merge support for DataFrame

### PR DESCRIPTION
**Pull request by hekaisheng**
## What do these changes do?

This PR implements a broadcast join method for DataFrames, when small DataFrame's chunk size is far less than bigger one, only shuffle the small one and broadcast them to the chunks of the other DataFrame, it will reduce time cost by avoiding shuffle of big DataFrame.

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them
- [x] #2774 merged

### CODE REVIEW

**Review by sighingnow — APPROVED** [review]

File: mars/dataframe/merge/merge.py [hunk 1]
```diff
@@ -246,15 +188,8 @@ def __call__(self, left, right):
...
-
-class DataFrameShuffleMerge(_DataFrameMergeBase):
-    _op_type_ = OperandDef.DATAFRAME_SHUFFLE_MERGE
-
-    def __init__(self, **kw):
-        super().__init__(**kw)
-
     @classmethod
-    def _gen_shuffle_chunks(cls, op, out_shape, shuffle_on, df):
+    def _gen_shuffle_chunks(cls, out_shape, shuffle_on, df):
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py:192]: Add type annotation.

File: mars/dataframe/merge/merge.py [hunk 2]
```diff
@@ -365,16 +300,8 @@ def _tile_one_chunk(cls, op, left, right):
         )
 
     @classmethod
-    def tile(cls, op):
+    def _tile_shuffle(cls, op, left, right):
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py:303]: Ditto.

File: mars/dataframe/merge/merge.py [hunk 3]
```diff
@@ -415,11 +342,161 @@ def tile(cls, op):
             columns_value=df.columns_value,
         )
 
+    @classmethod
+    def _tile_broadcast(cls, op, left, right):
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py:346]: Ditto.

File: mars/dataframe/merge/merge.py [hunk 4]
```diff
@@ -467,24 +544,192 @@ def _prepare_shuffle_on(use_index, side_on, on):
...
+        If specified, checks if merge is of specified type.
+
+        * "one_to_one" or "1:1": check if merge keys are unique in both
+          left and right datasets.
+        * "one_to_many" or "1:m": check if merge keys are unique in left
+          dataset.
+        * "many_to_one" or "m:1": check if merge keys are unique in right
+          dataset.
+        * "many_to_many" or "m:m": allowed, but does not result in checks.
+    strategy : {"shuffle", "broadcast"}, default None
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py:639]: For consistency, use `method` instead of `strategy`.

File: mars/dataframe/merge/merge.py [hunk 5]
```diff
@@ -496,14 +741,166 @@ def merge(
...
-    df, other, on=None, how="left", lsuffix="", rsuffix="", sort=False, strategy=None
-):
+    df: Union[DataFrame, Series],
+    other: Union[DataFrame, Series],
+    on: str = None,
+    how: str = "left",
+    lsuffix: str = "",
+    rsuffix: str = "",
+    sort: bool = False,
+    strategy: str = None,
```
Comment (qinxuye) [inline mars/dataframe/merge/merge.py:759]: Ditto.

**Review by qinxuye — COMMENTED** [review]
I left some comments.

**Review by wjsi — APPROVED** [review]
LGTM

**Review by qinxuye — APPROVED** [review]
LGTM

**Pull request merged**